examples/chol.dx

```diff
@@ -0,0 +1,106 @@
+' # Cholesky Factorization
+https://en.wikipedia.org/wiki/Cholesky_decomposition
+
+' ## Helpers
+
+ixkey2:: A n::Ix m::Ix. Key -> n -> m -> Key
+ixkey2 x i = ixkey $ ixkey x i
+
+' ### Index checking
+
+inUpper:: A m::Ix. m -> m -> Bool
+inUpper i j = (asint i) < (asint j)
+
+inLower:: A m::Ix. m -> m -> Bool
+inLower i j = (asint i) > (asint j)
+
+inDiag:: A m::Ix. m -> m -> Bool
+inDiag i j = (asint i) == (asint j)
+
+' ### Matrix Math
+
+vdot :: n=>Real -> n=>Real -> Real
+vdot x y = sum for i. x.i * y.i
+
+
+mmp :: l=>m=>Real -> m=>n=>Real -> l=>n=>Real
+mmp m1 m2 = for i k. sum for j. m1.i.j * m2.j.k
+
+
+eyeLike:: n=>n=>Real -> n=>n=>Real
+eyeLike x = for i. for j.
+  select (inDiag i j) 1.0 0.0
+
+transpose::n=>m=>Real->m=>n=>Real
+transpose x = for i. for j. x.j.i
+
+mmadd:: (n=>m=>Real)->(n=>m=>Real)->(n=>m=>Real)
+mmadd x y = for i. for j. x.i.j + y.i.j
+
+
+
+' ## Cholesky Algorithm
+
+chol :: (n=>n=>Real) -> (n=>n=>Real)
+chol x = ..
+  diagOp r j = 
+    innersum = sum for k::n. ..
+      select (inLower j k) (sq r.j.k) 0.0
+    sqrt (x.j.j - innersum)
+  lowerOp r i j = 
+    innersum = sum for k. ..
+      select (inLower j k) (r.i.k * r.j.k) 0.0
+    jdiag = diagOp r j -- can't look up r.j.j as it is not computed yet.
```

Comment:
this bit was the key thing we had wrong before.
We were reading the `r.j.j` before it is set to the correct value.
`r` is only done up to row `j-i` at this time.

So we need to recompute it here.
Hopefully dex can do Common Subexpression Elimnation and remove this.
But its still kinda ugly.